snuba/admin/clickhouse/migration_checks.py

```diff
@@ -0,0 +1,215 @@
+        return list(self.__migration_statuses.keys())
+
+    def can_run(self, migration_id: str) -> ResultReason:
+        """
+        Covers the following cases:
+        * no running a migration that is already pending or completed
+        * no running a migration that has a preceeding migration that is not completed
+        """
+        all_migration_ids = self.all_migration_ids
+        if self.__migration_statuses[migration_id]["status"] != Status.NOT_STARTED:
+            return ResultReason(False, Reason.ALREADY_RUN)
+
+        for m in all_migration_ids[: all_migration_ids.index(migration_id)]:
+            if self.__migration_statuses[m]["status"] != Status.COMPLETED:
+                return ResultReason(False, Reason.NEEDS_EARLIER_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        """
+        Covers the following cases:
+        * no reversing a migration that is has not started
+        * no reversing a migration that has a subsequent migration that has not been run
+        """
+        all_migration_ids = self.all_migration_ids
+        if self.__migration_statuses[migration_id]["status"] == Status.NOT_STARTED:
+            return ResultReason(False, Reason.NOT_RUN_YET)
+
+        for m in all_migration_ids[all_migration_ids.index(migration_id) + 1 :]:
+            if self.__migration_statuses[m]["status"] != Status.NOT_STARTED:
+                return ResultReason(False, Reason.NEEDS_SUBSEQUENT_MIGRATIONS)
+
+        return ResultReason(True, Reason.NO_REASON_NEEDED)
+
+
+class PolicyChecker(Checker):
+    """
+    The PolicyChecker validates whether you can run or
+    reverse a migration based on the policy for the
+    migration's group.
+
+    Policies are defined in the ADMIN_ALLOWED_MIGRATION_GROUPS
+    setting.
+    """
+
+    def __init__(self, migration_group: MigrationGroup) -> None:
+        self.__migration_group: MigrationGroup = migration_group
+        self.__policy: MigrationPolicy = get_migration_group_polices()[
+            migration_group.value
+        ]
+
+    def _get_migration_key(self, migration_id: str) -> MigrationKey:
+        return MigrationKey(self.__migration_group, migration_id)
+
+    def can_run(self, migration_id: str) -> ResultReason:
+        key = self._get_migration_key(migration_id)
+        if self.__policy.can_run(key):
+            return ResultReason(True, Reason.NO_REASON_NEEDED)
+        else:
+            return ResultReason(False, Reason.RUN_POLICY)
+
+    def can_reverse(self, migration_id: str) -> ResultReason:
+        key = self._get_migration_key(migration_id)
+        if self.__policy.can_reverse(key):
+            return ResultReason(True, Reason.NO_REASON_NEEDED)
+        else:
+            return ResultReason(False, Reason.REVERSE_POLICY)
+
+
+def do_checks(
+    checkers: Sequence[Checker], migration_id: str
+) -> Tuple[ResultReason, ResultReason]:
```

Comment:
Please make this a named tuple. 
Imagine you were reading the code that calls `do_checks`, that could be somewhere else in the code. How do you know which one of the field is the result for `run` and which one is the result for `reverse` ? 
Some body will eventually mess up the two while this can be prevented by having a named tuple that clearly states which one is which.

Reply:
this is updated now that the result types are different